File: NekoGirlStateMachine.py
import threading
from enum import Enum
from flask_socketio import emit
import structlog



class Emotion(Enum):
    NEUTRAL = "neutral"
    HAPPY = "happy"
    SAD = "sad"
    ANGRY = "angry"
    SURPRISED = "surprised"

class Animation(Enum):
    NONE = "none"
    DANCING = "silly_dancing"
    ANGRY = "angry"
    DAGGER = "dagger"

class NekoGirlStateMachine:
    def __init__(self, activity_timeout=30):
        # Initialize self.logger
        self.logger = structlog.getLogger(__name__)
        self.current_emotion = Emotion.NEUTRAL
        self.current_animation = Animation.NONE
        self.current_audio = None
        self.activity_timeout = activity_timeout
        self.reset_timer = None

    def set_emotion(self, emotion: Emotion):
        self.current_emotion = emotion
        self.emit('set_emotion', {'emotion': emotion.value})
        self.reset_activity_timer()
        self.logger.info(f"Emotion set to {self.current_emotion}")

    def set_animation(self, animation: Animation):
        self.current_animation = animation
        self.emit('set_animation', {'animation': animation.value})
        self.reset_activity_timer()
        self.logger.info(f"Animation set to {self.current_animation}")

    def set_audio(self, audio_url: str):
        self.current_audio = audio_url
        self.emit('set_audio', {'audio_url': audio_url})
        self.reset_activity_timer()
        self.logger.info(f"Audio URL set to {self.current_audio}")

    def reset_activity_timer(self):
        if self.reset_timer:
            self.reset_timer.cancel()
        self.reset_timer = threading.Timer(self.activity_timeout, self.reset_to_neutral)
        self.reset_timer.start()

    def reset_to_neutral(self):
        self.current_emotion = Emotion.NEUTRAL
        self.current_animation = Animation.NONE
        self.current_audio = None
        self.emit('state_reset', {'emotion': self.current_emotion, 'animation': self.current_animation})
        self.logger.info("Reset to neutral due to inactivity")

File: NekoGirlStateMachine.js
const EventEmitter = require('events');

class Emotion {
  static get NEUTRAL() {
    return 'neutral';
  }
  static get HAPPY() {
    return 'happy';
  }
  static get SAD() {
    return 'sad';
  }
  static get ANGRY() {
    return 'angry';
  }
  static get SURPRISED() {
    return 'surprised';
  }
}

class Animation {
  static get NONE() {
    return 'none';
  }
  static get DANCING() {
    return 'silly_dancing';
  }
  static get ANGRY() {
    return 'angry';
  }
  static get NEUTRAL() {
    return 'neutral';
  }
  static get LOOK_LEFT() {
    return 'look_left';
  }
  static get LOOK_RIGHT() {
    return 'look_right';
  }
}

class LookDirections {
  static get FORWARD() {
    return 'forward';
  }
  static get UP() {
    return 'up';
  }
  static get DOWN() {
    return 'down';
  }
  static get LEFT() {
    return 'left';
  }
  static get RIGHT() {
    return 'right';
  }
}

class NekoGirlStateMachine extends EventEmitter {
  constructor(activityTimeout = 30) {
    super();
    this.logger = console; // Replace with your preferred logger
    this.currentEmotion = Emotion.NEUTRAL;
    this.currentAnimation = Animation.NONE;
    this.currentAudio = null;
    this.activityTimeout = activityTimeout;
    this.resetTimer = null;
  }

  setEmotion(emotion) {
    this.currentEmotion = emotion;
    this.emit('emotion_change', { emotion });
    this.resetActivityTimer();
    this.logger.info(`Emotion set to ${this.currentEmotion}`);
  }

  setAnimation(animation) {
    this.currentAnimation = animation;
    this.emit('animation_change', { animation });
    this.resetActivityTimer();
    this.logger.info(`Animation set to ${this.currentAnimation}`);
  }

  setAudio(audioUrl) {
    this.currentAudio = audioUrl;
    this.emit('audio_change', { audioUrl });
    this.resetActivityTimer();
    this.logger.info(`Audio URL set to ${this.currentAudio}`);
  }

  resetActivityTimer() {
    if (this.resetTimer) {clearTimeout(this.resetTimer);}
    this.resetTimer = setTimeout(
      () => this.resetToNeutral(),
      this.activityTimeout * 1000,
    );
  }

  resetToNeutral() {
    this.currentEmotion = Emotion.NEUTRAL;
    this.currentAnimation = Animation.NONE;
    this.currentAudio = null;
    this.emit('state_reset', {
      emotion: this.currentEmotion,
      animation: this.currentAnimation,
    });
    this.logger.info('Reset to neutral due to inactivity');
  }
}

module.exports = { NekoGirlStateMachine, Emotion, Animation, LookDirections };
